src/hardware/hantek-dso/protocol.c

```diff
@@ -411,6 +595,80 @@ static int dso_set_voltage(const struct sr_dev_inst *sdi)
 	devc = sdi->priv;
 	usb = sdi->conn;
 
+	memset(cmdstring, 0, sizeof(cmdstring));
+	cmdstring[0] = CMD_SET_VOLTAGE;
+	/* TODO check voltage and relais settings */
+	cmdstring[2] = 0x08;
+
+	if (send_begin(sdi) != SR_OK)
+		return SR_ERR;
+
+	if ((ret = libusb_bulk_transfer(usb->devhdl, DSO_EP_OUT,
+			cmdstring, sizeof(cmdstring), &tmp, 100)) != 0) {
+		sr_err("Failed to set voltage: %s.", libusb_error_name(ret));
+		return SR_ERR;
+	}
+	sr_dbg("Sent CMD_SET_VOLTAGE.");
+
+
+	/* CH1 volts/div is encoded in bits 0-1 */
+	sr_dbg("CH1 vdiv index: %d.", devc->voltage[0]);
+	switch (devc->voltage[0]) {
+	case VDIV_1V:
+	case VDIV_100MV:
+	case VDIV_10MV:
+		cmdstring[2] |= 0x00;
+		break;
+	case VDIV_2V:
+	case VDIV_200MV:
+	case VDIV_20MV:
+		cmdstring[2] |= 0x01;
+		break;
+	case VDIV_5V:
+	case VDIV_500MV:
+	case VDIV_50MV:
+		cmdstring[2] |= 0x02;
+		break;
+	}
+
+	/* CH2 volts/div is encoded in bits 2-3 */
+	sr_dbg("CH2 vdiv index: %d.", devc->voltage[1]);
+	switch (devc->voltage[1]) {
+	case VDIV_1V:
+	case VDIV_100MV:
+	case VDIV_10MV:
+		cmdstring[2] |= 0x00;
+		break;
+	case VDIV_2V:
+	case VDIV_200MV:
+	case VDIV_20MV:
+		cmdstring[2] |= 0x04;
+		break;
+	case VDIV_5V:
+	case VDIV_500MV:
+	case VDIV_50MV:
+		cmdstring[2] |= 0x08;
+		break;
+	}
+
+	return SR_OK;
+}
+
+static int dso_set_voltage(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc;
+	struct sr_usb_dev_inst *usb;
+	int ret, tmp;
+	uint8_t cmdstring[8];
+
+	devc = sdi->priv;
+	if (devc->profile->fw_pid == 0x2250)
+		return dso2250_set_voltage(sdi);
```

Comment:
I've gotta say, I really feel like these should be setup as function pointers in the init stage or something, not sprinkled into every function, but maybe this is best for now.